// @vercel/postgres-kysely/dist/index.js
// src/index.ts
import {
  PostgresDialect,
  PostgresDriver,
  Kysely,
  PostgresAdapter
} from "kysely";
import { createPool } from "@vercel/postgres";

// src/error.ts
var VercelPostgresKyselyError = class extends Error {
  constructor(code, message) {
    super(`VercelPostgresError - '${code}': ${message}`);
    this.code = code;
    this.name = "VercelPostgresError";
  }
};

// src/index.ts
var VercelPostgresAdapter = class extends PostgresAdapter {
  constructor() {
    super(...arguments);
    // represented as readonly property to satisfy eslint rule:
    // typescript-eslint/class-literal-property-style
    this._supportsTransactionalDdl = false;
  }
  get supportsTransactionalDdl() {
    return this._supportsTransactionalDdl;
  }
};
var VercelPostgresDialect = class extends PostgresDialect {
  constructor(config) {
    super(config);
    this.config = config;
  }
  createAdapter() {
    return new VercelPostgresAdapter();
  }
  createDriver() {
    return new VercelPostgresPoolDriver(this.config);
  }
};
var VercelPostgresPoolDriver = class extends PostgresDriver {
  // Rather than trying to rebuild a perfectly good connection pool,
  // we can just use a proxy to throw if the user tries to stream.
  async acquireConnection() {
    const connection = await super.acquireConnection();
    return new Proxy(connection, {
      get(target, p) {
        const original = target[p];
        if (p === "streamQuery" && typeof original === "function") {
          throw new VercelPostgresKyselyError(
            "kysely_streaming_not_supported",
            "Streaming is not supported yet."
          );
        }
        if (typeof original === "function") {
          return original.bind(target);
        }
        return original;
      }
    });
  }
  beginTransaction() {
    throw new VercelPostgresKyselyError(
      "kysely_transactions_not_supported",
      "Transactions are not supported yet."
    );
  }
  commitTransaction() {
    throw new VercelPostgresKyselyError(
      "kysely_transactions_not_supported",
      "Transactions are not supported yet."
    );
  }
  rollbackTransaction() {
    throw new VercelPostgresKyselyError(
      "kysely_transactions_not_supported",
      "Transactions are not supported yet."
    );
  }
};
function createKysely(poolConfig, kyselyConfig) {
  return new Kysely({
    ...kyselyConfig,
    dialect: new VercelPostgresDialect({
      ...poolConfig,
      pool: createPool(poolConfig)
    })
  });
}
export {
  createKysely
};
